_source/_posts/2021-09-30-oauth-sessions-with-java.adoc

```diff
@@ -0,0 +1,316 @@
+
+toc::[]
+
+== Prerequisites
+
+- https://docs.docker.com/get-docker/[Docker 20.10+]
+- https://adoptopenjdk.net/?variant=openjdk11[Java 11+], use https://sdkman.io/[SDKMAN] to manage multiple Java versions!
+- https://httpie.io/download[HTTPie], a user-friendly HTTP CLI
+- https://cli.okta.com/[Okta CLI]
+
+== Sessions and OAuth 2.0 Applications
+
+An application that uses an OAuth 2.0 redirect, an https://oauth.net/2/grant-types/authorization-code/["Authorization Code Grant"], typically uses a server-side session to temporarily store the state of the login process until it completes. Let's look at a simplified OAuth login flow:
+
+image::{% asset_path 'blog/oauth-sessions-with-java/oauth-code-flow.png' %}[alt=Login Flow Sequence Diagram,width=800,align=center]
+
+1. User requests to log in - automatically when requesting a protected page or when the user clicks on a login button.
+2. Web App stores information about the current session - OAuth state info, and optionally a PKCE code verifier and/or a nonce for OpenId Connect (OIDC).
+3. Respond with a browser redirect to the authorization server.
+4. User interacts with the authorization server to provide credentials.
+5. Issue redirect with a code back to Web App.
+6. Finalize user authorization on the backend.
+7. The web app reads the previously stored OAuth data from the session.
+8. Verify the user has been authorized with the authorization server.
+9. Auth Server response with an OAuth access token.
+10. Web App stores access token in session.
+11. User is logged in.
+
+Two separate requests are made to the web application, the initial login request, and a verification step; both requests access the same session information. For your application, this means once you start scaling, you need to think about session management.
+
+== Stateless with JWTs
+
+At this point, you might be wondering if you could cram everything in a JWT and make this process stateless; removing the need for any session clustering. You _could_, but you would need to use a JWE (JSON Web Encryption) to ensure the browser didn't have access to any sensitive data. It is outside the scope of this post, but there are a whole host of reasons link:/blog/2017/08/17/why-jwts-suck-as-session-tokens[not to use JWTs as session tokens].
+
+== Load Balancer Sticky Sessions
+
+Another option to avoid session replication or clustering is to use "sticky sessions" (or "session affinity"), but this creates a fragile environment. If a web server fails or is taken out down for any reason, all users associated with that server would be essentially be logged out. Sticky sessions also violate the https://12factor.net/processes[12 Factor principles of process statelessness].
```

Comment:
```suggestion
Another option to avoid session replication or clustering is to use "sticky sessions" (or "session affinity"), but this creates a fragile environment. If a web server fails or is taken down for any reason, all users associated with that server would be essentially be logged out. Sticky sessions also violate the https://12factor.net/processes[12 factor principles of process statelessness].
```
duplicate verb-age